src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+        //}
+
+        private static bool IsConditional(IBinaryOperation operation)
+        {
+            switch (operation.OperatorKind)
+            {
+                case BinaryOperatorKind.ConditionalOr:
+                case BinaryOperatorKind.ConditionalAnd:
+                    return true;
+            }
+
+            return false;
+        }
+
+        public override IOperation VisitBinaryOperator(IBinaryOperation operation, object argument)
+        {
+            if (IsConditional(operation))
+            {
+                return VisitBinaryConditionalOperator(operation, true);
+            }
+
+            return base.VisitBinaryOperator(operation, argument);
+        }
+
+        public override IOperation VisitUnaryOperator(IUnaryOperation operation, object argument)
+        {
+            // PROTOTYPE(dataflow): ensure we properly detect logical Not
+            if (operation.OperatorKind == UnaryOperatorKind.Not)
+            {
+                return VisitConditionalExpression(operation.Operand, sense: false);
+            }
+
+            return base.VisitUnaryOperator(operation, argument);
+        }
+
+        private IOperation VisitBinaryConditionalOperator(IBinaryOperation binOp, bool sense)
+        {
+            bool andOrSense = sense;
+
+            switch (binOp.OperatorKind)
+            {
+                case BinaryOperatorKind.ConditionalOr:
+                    Debug.Assert(binOp.LeftOperand.Type.SpecialType == SpecialType.System_Boolean);
+                    Debug.Assert(binOp.RightOperand.Type.SpecialType == SpecialType.System_Boolean);
+
+                    // Rewrite (a || b) as ~(~a && ~b)
+                    andOrSense = !andOrSense;
+                    // Fall through
+                    goto case BinaryOperatorKind.ConditionalAnd;
+
+                case BinaryOperatorKind.ConditionalAnd:
+                    Debug.Assert(binOp.LeftOperand.Type.SpecialType == SpecialType.System_Boolean);
+                    Debug.Assert(binOp.RightOperand.Type.SpecialType == SpecialType.System_Boolean);
+
+                    // ~(a && b) is equivalent to (~a || ~b)
+                    if (!andOrSense)
+                    {
+                        // generate (~a || ~b)
+                        return VisitShortCircuitingOperator(binOp, sense, sense, true);
+                    }
+                    else
+                    {
+                        // generate (a && b)
+                        return VisitShortCircuitingOperator(binOp, sense, !sense, false);
+                    }
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(binOp.OperatorKind);
+            }
+        }
+
+        private IOperation VisitShortCircuitingOperator(IBinaryOperation condition, bool sense, bool stopSense, bool stopValue)
```

Comment:
>VisitShortCircuitingOperator [](start = 27, length = 28)

Since `stopSense` and `stopValue` are fully determined by two cases (AND or OR, as shown in table below), the arguments can be simplified to take `sense` and a flag (indicating AND vs. OR case).

Reply:
> Since stopSense and stopValue are fully determined by two cases (AND or OR, as shown in table below), the arguments can be simplified to take sense and a flag (indicating AND vs. OR case).

I don't see how is that going to be different, a flag is a flag and the more explicit it is (i.e. doesn't require additional interpretation), the better. Also, since we are pushing logical Not operations down to the leafs, the ```stopValue``` might not match the original operator for the node, which would create more confusion if we pass a flag indicating AND vs. OR, I believe.   

---
In reply to: [162208062](https://github.com/dotnet/roslyn/pull/24263#discussion_r162208062) [](ancestors = 162208062)

Reply:
The fact that we need a table to help interpretation is a strong indication that it's simpler to understand this method's contract from two cases (AND vs. OR), rather four possible cases (stopSense/stopValue permutations). 
That two of those permutations are illegal is another indication that one flag would be simpler, so it cannot be called with wrong arguments by accident.

---
In reply to: [162214164](https://github.com/dotnet/roslyn/pull/24263#discussion_r162214164) [](ancestors = 162214164,162208062)

Reply:
From discussion with @AlekseyTs, the key that I missed is that the two calls to `VisitShortCircuitingOperator` can both process AND and OR operations (the comments make it look like one is for AND and the other is for OR).
The permutations are actually legal.
Thanks for the clarification.